README.md

```diff
@@ -1 +1,189 @@
+{
+  "role": "user",  // Parsed by an IPLD implementation
+  "role": "admin", // Malicious duplicate field, omitted by the IPLD parser, accpeted by the browser
+  "links": [
+    {"/": "bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"},
+    {"/": "bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"},
+    {"/": "bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"}
+  ],
+  "sig": "8ufaS9w3CGN8cbQTUSoL1i7eaKiWLSXsD2LbZVmvM9zF"
+}
+```
+
+In the above example, the canonicalization step MAY lead to the signature validating, but the client parsing the `role: "admin"` field instead.
+
+# 2 Safety
+
+Data that has already been parsed to an in-memory IPLD representation can be canonically encoded to a trivially: it has already been through a [parser / validator](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).
+
+Data purporting to conform to an IPLD encoding (such as [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/)) MUST be validated prior to signature verification. This can be as simple as round-trip decoding/encoding the JSON and checking that the hash matches. A validation error MUST be signalled if it does not match. 
+
+> [Implementers] may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
+>
+> — [DAG-JSON Spec](https://ipld.io/specs/codecs/dag-json/spec/)
+
+As it is critical for signatures guard against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
+
+## 2.2 Raw (Noncanonicalized) Data
+
+Canonicalization is not required if data is encoded as raw bytes (multicodec `0x00`). The exact bytes are already present, and MUST not be changed.
+
+# 3 Varsig Format
+
+A varsig is a bytestring that includes the following information:
+
+| Segment Name              | Type     | Description                                     | Required |
+|---------------------------|----------|-------------------------------------------------|----------|
+| Varsig Prefix             | `0xd000` | The multicodec varsig prefix                    | Yes      |
+| Public Key Prefix         | `Varint` | The multicodec prefix for the public key type   | Yes      |
+| Hash Prefix               | `Varint` | The multicodec prefix for the hash algorithm    | Yes      |
+| Hash Length               | `Varint` | The hash length                                 | Yes      |
+| Content Multicodec Prefix | `Varint` | The IPLD encoding uses to canonicalize the data | Yes      |
+| Raw Signature             | `Varint` | The raw signature                               | Yes      |
```

Comment:
In ucan-ipld we have [defined `0xd000` as `NonStandardSignature`](https://github.com/ucan-wg/ucan-ipld/#251-nonstandard-signatures) which had this layout

```ts
<varint 0xd000><varint sig_size><bytes sig><utf8 alg>
```

Last bit `<utf8 alg>` was important to enable formatting UCANs into JWT, because there is no other way to tell what should go into JWT header algorithm field. This spec no longer has this which creates problem for UCAN codec as it would need to know all possible algorithm names even non standard ones or fail.

I think we should retain [NonStandard Signature](https://github.com/ucan-wg/ucan-ipld/#251-nonstandard-signatures) idea here allowing known algorithms to be compact, while still supporting arbitrary ones at expense of extra bytes.

Reply:
Good point: changed (proposed) varsig code to `0x34` 👍

Reply:
> Last bit `<utf8 alg>` was important to enable formatting UCANs into JWT

Hmm yeah. JWT assumes that you're signing over a base64 string. To make sure that I'm understanding you correctly: you're saying to use `0xd000 Nonstandard Signature` to say "the application knows how to encode and check this one, but some fancy encoding needs to happen first"?

Reply:
> Hmm yeah. JWT assumes that you're signing over a base64 string. 

This is unrelated

> To make sure that I'm understanding you correctly: you're saying to use 0xd000 Nonstandard Signature to say "the application knows how to encode and check this one, but some encoding needs to happen first that is outside of the varsig system"?

What I'm saying is that we either:

1. Have a table for translating all the varsig prefixes into string labels like EdDSA, RS256 etc...
2. Allow encoding label into varsig itself

In dag-ucan we went for the hybrid where we have:
  1. A table for standard ones so we can translate between varints to names they correspond
  2. A non-standard signature encoding that allows encoding label into a signature so you don't have to be in a standard table

This allows new signing algorithms to be deployed without having to go through standards table inclusion process and then getting all the software to incorporate it. This way all the software that understands varsig is able to display the signature name and rendering them in places like JWT headers `alg`.

More concretely dag-ucan can encoder / decoder roundtrip between JWT and IPLD representations because even non-standard algorithms will have label in them.

Alternative for dag-ucan would be to take that label and hoist it into some other field, however it makes sense to include it in the signature itself because translation between varint parameters and some name is generally useful in many other contexts. There is also 0 overhead for standard signatures so nothing to loose there. This way things can start as non standard and have good experience and shave some bytes after they're standardize and get adoption

Reply:
TL;DR

If you're suggesting that we leave this all at the schema layer and let the application figure out how to get the data into and out of the right signing format, then I agree with you. We just need to be clear about that the exact scope of Varsig itself is.

> In dag-ucan we went for the hybrid

Right, but DAG-UCAN was never merged into Multicodec. IIRC the reason was that it can be described with existing DAG-* methods.

> More concretely dag-ucan can encoder / decoder roundtrip between JWT and IPLD representations because even non-standard algorithms will have label in them.

How do you signal that something is in DAG-UCAN? My understanding is that you know that something conforms to the `ucan-ipld` schema from the application only.

> This is unrelated

It _is_ related, though: if you want to format a UCAN to JWT using varsig, it needs to know that it's being put into base64 before hashing. Otherwise verification will fail. Passing around some UCAN-IPLD will need to know this information. There is no DAG-UCAN, so we need to either base64 serialize the payload, add this info to the varsig, or have the application infer this information from context. We need to be clear about which strategy we're taking.

Reply:
I've been thinking about how to properly encode JOSE signature in varsig. This should be relevant to encoding UCAN, and I think the discussion above.

First of all this would be simple for JWS but more tricky for JWT. Reason being that JWTs contain a `typ` property in the protected header set to `"JWT"`. There are also a bunch of other fields that can be put in the protected header. Here's a suggestion for how a varsig codec for JOSE could work: 

1. Load the payload IPLD object 
2. Check for a `_protected` field in (1), here additional protected header fields can be added
3. Create the real protected header by combining (2) with the alg from `key_multicodec`
4. Encode the rest of the IPLD object with `dag-json`
5. Combine the two steps above `<base64url(3)>.<base64url(4)>`
6. Sign, or verify signature


This would need to be a new `dag-jose-varsig`(or ideally an upgrade to `dag-jose`)

Reply:
> It is related, though: if you want to format a UCAN to JWT using varsig, it needs to know that it's being put into base64 before hashing. Otherwise verification will fail. 

I think you’ve misunderstood what I was commenting on. You seem to be talking about capturing information how to go from data to bytes that will be signed. I agree we do need to capture that and that is why I suggested making format a multicodec as opposed to a varint that way it can nest as much info as needed inside it.

However that that is not what I was talking about. What I was talking about is that this proposal no longer allows has a section to put a name that can be rendered, which is something we lost from previous iteration and I was trying to explain how it is useful and why I wish us to keep it.

Reply:
> If you're suggesting that we leave this all at the schema layer and let the application figure out how to get the data into and out of the right signing format

I’m not suggesting this